pilot/pkg/model/network_gw_names.go

```diff
@@ -0,0 +1,128 @@
+package model
+
+import (
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type nameCacheEntry struct {
+	value  []string
+	expiry time.Time
+}
+
+type networkGatewayNameCache struct {
+	sync.Mutex
+	client     *dnsClient
+	cache      map[string]nameCacheEntry
+	xdsUpdater XDSUpdater
+}
+
+func newNetworkGatewayNameCache(xdsUpdater XDSUpdater) (*networkGatewayNameCache, error) {
+	c, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+	return &networkGatewayNameCache{client: c, xdsUpdater: xdsUpdater, cache: map[string]nameCacheEntry{}}, nil
+}
+
+func (n *networkGatewayNameCache) Resolve(name string) []string {
+	n.Lock()
+	defer n.Unlock()
+	if entry, ok := n.cache[name]; ok && entry.expiry.After(time.Now()) {
+		return entry.value
+	}
+	return n.resolveAndCache(name)
+}
+
+func (n *networkGatewayNameCache) resolveAndCache(name string) []string {
+	delete(n.cache, name)
+	names, ttl := n.resolve(name)
+	// don't cache for very small TTL or empty list of names
+	if len(names) == 0 || ttl < 10*time.Second {
+		return names
+	}
+	expiry := time.Now().Add(ttl)
+	time.AfterFunc(ttl, func() {
+		log.Infof("triggering update due to gateway hostname TTL expiry")
+		n.xdsUpdater.ConfigUpdate(&PushRequest{Full: true, Reason: []TriggerReason{NetworksTrigger}})
+	})
+	n.cache[name] = nameCacheEntry{
+		value:  names,
+		expiry: expiry,
+	}
+	return names
+}
+
+// resolve gets all the A and AAAA records for the given name
+func (n *networkGatewayNameCache) resolve(name string) ([]string, time.Duration) {
+	res := n.client.Query(new(dns.Msg).SetQuestion(name, dns.TypeA|dns.TypeAAAA))
+	if res == nil || len(res.Answer) == 0 {
+		return nil, 0
+	}
+	ttl := uint32(math.MaxUint32)
+	var out []string
+	for _, rr := range res.Answer {
+		switch v := rr.(type) {
+		case *dns.A:
+			out = append(out, v.A.String())
+		case *dns.AAAA:
+			out = append(out, v.AAAA.String())
+		default:
+			// not a valid record, don't inspect TTL
+			continue
+		}
+		if nextTTL := rr.Header().Ttl; nextTTL < ttl {
+			ttl = nextTTL
+		}
+	}
+	return nil, time.Duration(ttl)
+}
+
+// TODO share code with pkg/dns
+type dnsClient struct {
+	*dns.Client
+	resolvConfServers []string
+}
+
+func newClient() (*dnsClient, error) {
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+	c := &dnsClient{
```

Comment:
Any reason to use miekg vs `net.Resolver.LookupIP`?

Reply:
I think the former is much lower level

Reply:
Ah but then we don't get the TTL. https://github.com/golang/go/issues/24796.

https://github.com/rs/dnscache does add this, but its still not using TTL so doesn't seem ideal

Seems like there ought to be a better way but I don't see any